pymc/sampling.py

```diff
@@ -1534,6 +1542,157 @@ def stop_tuning(step):
+    for node in nodes:
+        if isinstance(node, SharedVariable) or node in fg.outputs or node in givens_dict:
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Propagate the volatile property to the node's clients
+        if is_volatile:
+            try:
+                clients = [c for c, _ in fg.clients[node]]  # type: ignore
+            except KeyError:
+                # Some Apply nodes don't show up in the clients of a FunctionGraph
+                # so we take the Apply node's outputs instead
+                clients = node.outputs  # type: ignore
+            for client in clients:
+                if (
+                    isinstance(client, Variable)
+                    and client.owner is not None
+                    and isinstance(client.owner.op, RandomVariable)
+                    and any(
+                        rv_param in volatile_nodes
+                        for rv_param in client.owner.inputs
+                        if not isinstance(
+                            rv_param, (RandomStateSharedVariable, RandomGeneratorSharedVariable)
+                        )
+                    )
+                ):
+                    # Random variables that have volatile parameters are also volatile
+                    volatile_nodes.add(client)
+                elif not (
+                    client in vars_in_trace
+                    and isinstance(client, Variable)
+                    and client.owner is not None
+                    and (
+                        isinstance(client.owner.op, RandomVariable)
+                        or getattr(client.tag, "symbolic_distribution", False)
+                    )
+                ):
+                    # Other variables that aren't random variables in the trace are volatile
+                    volatile_nodes.add(client)
+
+    # Collect the function inputs by walking the graph from the outputs. Inputs will be:
+    # 1. Random variables that are not volatile
+    # 2. Variables that have no owner and are not constant or shared
+    inputs = []
+
+    def expand(node):
+        if isinstance(node, Apply):
+            # Apply nodes are never inputs
+            return node.get_parents()
+        assert isinstance(node, Variable)
+        if (
+            (
+                node.owner is None and not isinstance(node, (Constant, SharedVariable))
+            )  # Variables without owners that are not constant or shared
+            or (
+                node.owner is not None and isinstance(node.owner.op, RandomVariable)
+            )  # Random variables
+            or (node in vars_in_trace)  # Variables in the trace
+        ):
+            if node not in volatile_nodes:
+                # This test will include variables without owners, and that are not constant
+                # or shared, because these nodes will never be considered volatile
+                inputs.append(node)
+        return node.get_parents()
```

Comment:
This seems to pass all tests locally:
```suggestion
    def expand(node):
        if (
            (
                (
                    node.owner is None
                    and not isinstance(node, (Constant, SharedVariable))
                )  # Variables without owners that are not constant or shared
                or node in vars_in_trace  # Variables in the trace
            )
            and node not in volatile_nodes
        ):
                # This test will include variables without owners, and that are not constant
                # or shared, because these nodes will never be considered volatile
                inputs.append(node)

        if node.owner:
            return node.owner.inputs
```